File: Dam-Control-System/src/main/java/com/SmartDam/Dam_Control_System/service/WeatherForecastService.java
package com.SmartDam.Dam_Control_System.service;

import com.fasterxml.jackson.databind.JsonNode;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;
import org.springframework.web.client.RestClient;
import java.util.List;

@Service
public class WeatherForecastService {

    private static final Logger log = LoggerFactory.getLogger(WeatherForecastService.class);
    private final RestClient restClient;

    public WeatherForecastService() {
        this.restClient = RestClient.builder()
                .baseUrl("https://api.open-meteo.com/v1/forecast")
                .build();
    }

    /**
     * Fetches 3-day precipitation forecast for dynamic coordinates.
     *
     * @return Sum of forecast precipitation in millimeters (mm) over the next 3 days.
     */
    public double fetchThreeDayPrecipitation(double latitude, double longitude) {
        try {
            log.info("Fetching 3-day weather forecast from Open-Meteo API for coordinates: Lat {}, Lon {}...", latitude, longitude);
            JsonNode response = restClient.get()
                    .uri(uriBuilder -> uriBuilder
                            .queryParam("latitude", latitude)
                            .queryParam("longitude", longitude)
                            .queryParam("daily", "precipitation_sum")
                            .queryParam("forecast_days", 3)
                            .queryParam("timezone", "auto")
                            .build())
                    .retrieve()
                    .body(JsonNode.class);

            if (response != null && response.has("daily") && response.get("daily").has("precipitation_sum")) {
                JsonNode precipSumNode = response.get("daily").get("precipitation_sum");
                double totalPrecipitation = 0.0;
                if (precipSumNode.isArray()) {
                    for (JsonNode valueNode : precipSumNode) {
                        if (!valueNode.isNull()) {
                            totalPrecipitation += valueNode.asDouble();
                        }
                    }
                }
                log.info("Successfully fetched 3-day weather forecast. Total Precipitation: {} mm", totalPrecipitation);
                return totalPrecipitation;
            } else {
                log.warn("Weather API response did not contain daily precipitation fields. Defaulting to 0.0 mm");
                return 0.0;
            }
        } catch (Exception e) {
            log.error("Failed to retrieve weather forecast from Open-Meteo: {}. Defaulting to 0.0 mm", e.getMessage(), e);
            return 0.0;
        }
    }

    public double fetchThreeDayPrecipitation() {
        return fetchThreeDayPrecipitation(20.1677, 79.3048);
    }
}

File: src/DamControlSystem/Services/WeatherForecastService.cs
using System.Globalization;
using System.Text.Json;

namespace DamControlSystem.Services;

public class WeatherForecastService : IWeatherForecastService
{
    private readonly HttpClient _httpClient;
    private readonly ILogger<WeatherForecastService> _logger;

    public WeatherForecastService(HttpClient httpClient, ILogger<WeatherForecastService> logger)
    {
        _httpClient = httpClient;
        _logger = logger;
    }

    public async Task<double> FetchThreeDayPrecipitationAsync(double latitude, double longitude)
    {
        try
        {
            _logger.LogInformation("Fetching 3-day weather forecast from Open-Meteo API for coordinates: Lat {Lat}, Lon {Lon}...", latitude, longitude);

            var latStr = latitude.ToString(CultureInfo.InvariantCulture);
            var lonStr = longitude.ToString(CultureInfo.InvariantCulture);
            var url = $"v1/forecast?latitude={latStr}&longitude={lonStr}&daily=precipitation_sum&forecast_days=3&timezone=auto";

            var response = await _httpClient.GetAsync(url);
            if (!response.IsSuccessStatusCode)
            {
                _logger.LogWarning("Open-Meteo API returned non-success code {StatusCode}. Defaulting to 0.0 mm", response.StatusCode);
                return 0.0;
            }

            using var document = await JsonDocument.ParseAsync(await response.Content.ReadAsStreamAsync());
            var root = document.RootElement;

            if (root.TryGetProperty("daily", out var daily) &&
                daily.TryGetProperty("precipitation_sum", out var precipArray) &&
                precipArray.ValueKind == JsonValueKind.Array)
            {
                double totalPrecipitation = 0.0;
                foreach (var item in precipArray.EnumerateArray())
                {
                    if (item.ValueKind == JsonValueKind.Number)
                    {
                        totalPrecipitation += item.GetDouble();
                    }
                }

                _logger.LogInformation("Successfully fetched 3-day weather forecast. Total Precipitation: {Total} mm", totalPrecipitation);
                return totalPrecipitation;
            }

            _logger.LogWarning("Weather API response did not contain daily precipitation fields. Defaulting to 0.0 mm");
            return 0.0;
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Failed to retrieve weather forecast from Open-Meteo: {Message}. Defaulting to 0.0 mm", ex.Message);
            return 0.0;
        }
    }

    public Task<double> FetchThreeDayPrecipitationAsync()
    {
        return FetchThreeDayPrecipitationAsync(20.1677, 79.3048);
    }
}
